rivertile/riverctl.zig

```diff
@@ -0,0 +1,81 @@
+const std = @import("std");
+const mem = std.mem;
+const io = std.io;
+const assert = std.debug.assert;
+const flags = @import("flags");
+const wayland = @import("wayland");
+const wl = wayland.client.wl;
+const zriver = wayland.client.zriver;
+
+const RiverControlError = error{
+    CannotGetRegistry,
+    RunCommandFailed,
+};
+
+const Context = struct {
+    control: ?*zriver.ControlV1 = null,
+    seat: ?*wl.Seat = null,
+};
+
+pub fn run(args: []const [*:0]const u8) !void {
+    const display = try wl.Display.connect(null);
+    defer display.disconnect();
+
+    const registry = display.getRegistry() catch {
+        return RiverControlError.CannotGetRegistry;
+    };
+    defer registry.destroy();
+
+    var ctx = Context{};
+
+    registry.setListener(*Context, registryListener, &ctx);
+    if (display.roundtrip() != .SUCCESS) std.log.err("initial roundtrip failed", .{});
+
+    const control = ctx.control orelse return error.RiverControlNotAdvertised;
+    const seat = ctx.seat orelse return error.SeatNotAdverstised;
+
+    for (args) |arg| control.addArgument(arg);
+
+    const callback = control.runCommand(seat) catch {
+        return RiverControlError.RunCommandFailed;
+    };
+    callback.setListener(?*anyopaque, callbackListener, null);
+
+    // Loop until our callback is called and we exit.
+    while (display.dispatch() != .SUCCESS) {
+        std.log.err("failed to dispatch wayland events", .{});
+    }
+}
+
+fn registryListener(registry: *wl.Registry, event: wl.Registry.Event, ctx: *Context) void {
+    switch (event) {
+        .global => |global| {
+            if (mem.orderZ(u8, global.interface, wl.Seat.getInterface().name) == .eq) {
+                assert(ctx.seat == null); // TODO: support multiple seats
+                ctx.seat = registry.bind(global.name, wl.Seat, 1) catch @panic("out of memory");
+            } else if (mem.orderZ(u8, global.interface, zriver.ControlV1.getInterface().name) == .eq) {
+                ctx.control = registry.bind(global.name, zriver.ControlV1, 1) catch @panic("out of memory");
+            }
+        },
+        .global_remove => {},
+    }
+}
+
+fn callbackListener(_: *zriver.CommandCallbackV1, event: zriver.CommandCallbackV1.Event, _: ?*anyopaque) void {
+    switch (event) {
+        .success => |success| {
+            if (mem.len(success.output) > 0) {
+                const stdout = io.getStdOut().writer();
+                stdout.print("{s}\n", .{success.output}) catch @panic("failed to write to stdout");
+            }
+        },
+        .failure => |failure| {
+            // A small hack to provide usage text when river reports an unknown command.
+            if (mem.orderZ(u8, failure.failure_message, "unknown command") == .eq) {
+                std.log.err("unknown command", .{});
+                return;
+            }
+            std.log.err("{s}", .{failure.failure_message});
+        },
+    }
+}
```

Comment:
In the unlikely event that Isaac wants this feature in rivertile, this would be the wrong way to implement it.

Why create a new wayland connection when we can just use the one we already have for the layout protocol?